Zend/zend_object_handlers.c

```diff
@@ -739,6 +742,12 @@ ZEND_API zval *zend_std_read_property(zval *object, zval *member, int type, void
 			} else {
 				retval = &EG(uninitialized_zval);
 			}
+
+			if (UNEXPECTED(ZEND_CLASS_HAS_TYPE_HINTS(zobj->ce) &&
+				(prop_info = zend_object_fetch_property_type_info(Z_OBJCE_P(object), name, cache_slot)))) {
```

Comment:
zend_object_fetch_property_type_info assumes that the passed cache_slot is valid. However this is not necessary true for the WRONG_PROPERTY_OFFSET case, as it does not initialize the cache slot. Repro:

```php
class A {
    public ?object $prop = null;
}

class B {
    private int $prop;

    public function __get($name) {
        return "foo";
    }
}

function test($obj) {
    var_dump($obj->prop);
}

test(new A);
test(new B);
```

```
NULL

Fatal error: Uncaught TypeError: Typed property A::$prop must be object or null, string used in /home/nikic/php-src-refs/t062.php:16
Stack trace:
#0 /home/nikic/php-src-refs/t062.php(20): test(Object(B))
#1 {main}
  thrown in /home/nikic/php-src-refs/t062.php on line 16
```

Reply:
Is `IS_WRONG_RPOPERTY_OFFSET(property_offset) ? NULL : cache_slot` good enough here or are there still edge cases?

Reply:
We also don't clear cache_slot on a non-static access to a static property (which falls back to a dynamic property access). I'd imagine the same issue could occur there.

It would be nicer if we could also get the property_info from the get_property_offset call -- not sure what the performance implications there would be...